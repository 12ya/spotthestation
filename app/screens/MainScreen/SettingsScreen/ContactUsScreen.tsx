import { useNavigation } from "@react-navigation/native"
import { observer } from "mobx-react-lite"
import React, { useCallback, useState } from "react"
import { ViewStyle, TextStyle, ScrollView, Pressable } from "react-native"
import { Dropdown } from "react-native-element-dropdown"
import email from 'react-native-email'
import { useSafeAreaInsets } from "react-native-safe-area-context"
import { Button, Icon, Screen, Text, TextField } from "../../../components"
import { colors, spacing, typography } from "../../../theme"
import { translate } from "../../../i18n"
import Config from "../../../config"


export const ContactUsScreen = observer(function ContactUsScreen() {
  const navigation = useNavigation()
  const topInset = useSafeAreaInsets().top
  const [title, setTitle] = useState('')
  const [comments, setComments] = useState('')
  const [ideas, setIdeas] = useState('')

  const $headerStyleOverride: TextStyle = {
    top: topInset + 24,
  }

  const handleSend = useCallback(() => {
    email(Config.CONTACT_EMAIL, {
      subject: title,
      body: `Comments:\n\n\t${comments}\n\nImprovement Ideas:\n\n\t${ideas}`,
      checkCanOpen: false
    }).catch(console.error)
  }, [title, comments, ideas])

  return (
    <Screen
      preset="fixed" 
      contentContainerStyle={[$container, $headerStyleOverride]} 
      style={{backgroundColor: colors.palette.neutral900}} 
      statusBarStyle="light"
    >
      <ScrollView
        accessible
        accessibilityLabel="Contact scrollable us area"
        accessibilityHint="Contact scrollable us area"
        accessibilityRole="scrollbar"
        style={$scrollContainer} 
        scrollEnabled 
        contentContainerStyle={$scrollSontentContainerStyle}
      >
        <Pressable 
          accessible
          accessibilityLabel="Back button"
          accessibilityHint="Navigates to the previous screen"
          accessibilityRole="button"
          onPress={() => navigation.goBack()} 
          style={$backButton}
        >
          <Icon icon="caretLeft" color={colors.palette.neutral250} />
          <Text tx="settings.contactUsData.backButton" style={$backButtonText} />
        </Pressable>
        <Text
          accessible
          accessibilityLabel="title"
          accessibilityHint="title"
          accessibilityRole="text"
          tx="settings.contactUsData.title" 
          style={$header} 
        />
        <Dropdown
          accessibilityLabel="title select"
          style={[$dropdown, $inputMargin]}
          placeholderStyle={[$dropdownText, $dropdownPlaceholder]}
          selectedTextStyle={[$dropdownText, $dropdownSelected]}
          placeholder={translate("settings.contactUsData.titlePlaceholder")}
          data={[
            { label: "App Crashing", value: "App Crashing" },
            { label: "Feature Suggestions", value: "Feature Suggestions" },
            { label: "General Questions", value: "General Questions" }
          ]}
          itemContainerStyle={{
            backgroundColor: colors.palette.neutral350,
          }}
          containerStyle={$dropdownContainer}
          itemTextStyle={$dropdownText}
          activeColor={colors.palette.neutral450}
          value={title}
          labelField="label"
          valueField="value"
          onChange={({ value }) => {
            setTitle(value as string)
          }}
          renderRightIcon={() => (
            <Icon
              icon="chevronDown"
              size={28}
              color={colors.palette.neutral450}
              containerStyle={$dropdownRightAccessory}
            />
          )}
        />
        <TextField
          accessible
          accessibilityLabel="comments input"
          accessibilityHint="comments input"
          accessibilityRole="text"
          value={comments}
          multiline
          numberOfLines={12}
          textAlignVertical="top"
          placeholderTx="settings.contactUsData.commentsPlaceholder"
          inputWrapperStyle={[$inputMargin, $multiline]}
          style={[$multiline, $inputWithoutPadding]}
          onChangeText={setComments}
        />
        <TextField
          accessible
          accessibilityLabel="ideas input"
          accessibilityHint="ideas input"
          accessibilityRole="text"
          value={ideas}
          multiline
          numberOfLines={12}
          textAlignVertical="top"
          placeholderTx="settings.contactUsData.ideasPlaceholder"
          inputWrapperStyle={[$inputMargin, $multiline]}
          style={[$multiline, $inputWithoutPadding]}
          onChangeText={setIdeas}
        />
        <Button
          accessible
          accessibilityLabel="send button"
          accessibilityHint="Navigates to the mail app"
          tx="settings.contactUsData.sendButton"
          textStyle={$buttonText}
          style={!title || !comments || !ideas ? [$button, $disabled] : $button}
          pressedStyle={$button}
          onPress={handleSend}
          disabled={!title || !comments || !ideas}
        />
      </ScrollView>
    </Screen>
  )
})

const $container: ViewStyle = {
  flex: 1,
  backgroundColor: colors.backgroundDark,
  height: '100%'
}

const $backButton: ViewStyle = {
  flexDirection: "row",
  alignItems: 'center',
  paddingBottom: 11
}

const $scrollSontentContainerStyle: ViewStyle = { 
  flexGrow: 1,
  paddingBottom: 60
}

const $scrollContainer: ViewStyle = { 
  paddingHorizontal: 18,
}

const $text: TextStyle = {
  fontFamily: typography.primary?.normal,
  fontSize: 18,
  lineHeight: 22,
  color: colors.palette.neutral450,
  textAlign: 'left',
  paddingBottom: 24
}

const $backButtonText: TextStyle = {
  ...$text,
  color: colors.palette.neutral250,
  paddingBottom: 0,
  paddingLeft: 5
}

const $header: TextStyle = {
  fontFamily: typography.primary.normal,
  fontSize: 36,
  lineHeight: 44,
  color: colors.palette.neutral250
}

const $dropdown: ViewStyle = {
  borderRadius: 28,
  height: 56,
  backgroundColor: colors.palette.neutral350,
  overflow: "hidden",
}

const $inputMargin: ViewStyle = {
  marginTop: 18
}

const $inputWithoutPadding: ViewStyle = {
  paddingTop: 0
}

const $disabled: ViewStyle = {
  opacity: .5
}

const $multiline: TextStyle = {
  height: 'auto',
  textAlignVertical: 'top',
  paddingTop: 10
}

const $dropdownContainer: ViewStyle = {
  backgroundColor: colors.palette.neutral350,
  borderRadius: 10,
  marginTop: -40,
  borderWidth: 0
}

const $dropdownPlaceholder: TextStyle = {
  color: colors.palette.neutral450,
}

const $dropdownSelected: TextStyle = {
  color: colors.palette.neutral250,
}

const $dropdownText: TextStyle = {
  flex: 1,
  // alignSelf: "stretch",
  fontFamily: typography.primary.normal,
  fontSize: 18,
  paddingVertical: 0,
  paddingHorizontal: 0,
  marginHorizontal: spacing.small,
  textAlignVertical: "center"
}

const $dropdownRightAccessory: ViewStyle = {
  marginEnd: spacing.large,
  height: 56,
  justifyContent: "center",
  alignItems: "center",
}

const $button: ViewStyle = {
  width: "100%",
  height: 64,
  backgroundColor: colors.palette.buttonBlue,
  borderRadius: 28,
  borderWidth: 0,
  marginVertical: 24
}

const $buttonText: TextStyle = {
  color: colors.palette.neutral100,
  fontSize: 18,
  fontFamily: typography.primary.medium,
  lineHeight: 21,
}
