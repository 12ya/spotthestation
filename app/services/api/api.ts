/**
 * This Api class lets you define an API endpoint and methods to request
 * data and process it.
 *
 * See the [Backend API Integration](https://github.com/infinitered/ignite/blob/master/docs/Backend-API-Integration.md)
 * documentation for more details.
 */
import {
  ApiResponse, // @demo remove-current-line
  ApisauceInstance,
  create,
} from "apisauce"
import Config from "../../config"
import { GeneralApiProblem, getGeneralApiProblem } from "./apiProblem" // @demo remove-current-line
import type {
  ApiConfig, GetISSSightingsParams, ISSSighting, ISSSightingResponse,
} from "./api.types"
import { GooglePlaceDetail } from "react-native-google-places-autocomplete"
import { ISSData } from "../../utils/loadAndFormatISSDate"
import { TimeZoneDataResponse } from "../../utils/geolocation"

/**
 * Configuring the apisauce instance.
 */
export const DEFAULT_API_CONFIG: ApiConfig = {
  url: Config.API_URL,
  timeout: 100000,
}

/**
 * Manages all requests to the API. You can use this class to build out
 * various requests that you need to call from your backend API.
 */
export class Api {
  apisauce: ApisauceInstance
  config: ApiConfig

  /**
   * Set up our API instance. Keep this lightweight!
   */
  constructor(config: ApiConfig = DEFAULT_API_CONFIG) {
    this.config = config
    this.apisauce = create({
      baseURL: this.config.url,
      timeout: this.config.timeout,
      headers: {
        Accept: "application/json",
      },
    })
  }

  async getPlaces(url: string, key: "results" | "candidates"): Promise<{ kind: "ok"; places: GooglePlaceDetail[] } | GeneralApiProblem> {
    const response: ApiResponse<Record<typeof key, GooglePlaceDetail[]>> = await this.apisauce.get(url, {}, { baseURL: "" })

    if (!response.ok) {
      const problem = getGeneralApiProblem(response)
      if (problem) return problem
    }

    return { kind: "ok", places: response.data[key] }
  }

  async getLocationTimeZone(url: string): Promise<TimeZoneDataResponse | GeneralApiProblem> {
    const response: ApiResponse<any> = await this.apisauce.get(url, {}, { baseURL: "" })

    if (!response.ok) {
      const problem = getGeneralApiProblem(response)
      if (problem) return problem
    }

    return { kind: "ok", zone: response.data }
  }

  async getISSData(): Promise<ISSData | GeneralApiProblem> {
    const response: ApiResponse<ISSData> = await this.apisauce.get("/tracking/nasa", {}, { baseURL: Config.ISS_TRAJECTORY_DATA_API_URL })
    
    if (!response.ok) {
      const problem = getGeneralApiProblem(response)
      if (problem) return problem
    }

    return response.data
  }

  async getISSSightings({ zone, lat, lon }: GetISSSightingsParams): Promise<ISSSightingResponse> {
    const response: ApiResponse<ISSSighting[]> = await this.apisauce.get(
      `/tracking/oem-nasa?zone=${zone}&lat=${lat}&lon=${lon}`, 
      {}, 
      { baseURL: Config.ISS_TRAJECTORY_DATA_API_URL }
    )
    
    if (!response.ok) {
      const problem = getGeneralApiProblem(response)
      if (problem) return { ok: false, data: response.data }
    }

    return { ok: true, data: response.data }
  }

  async getLocationByCoords(url: string): Promise<any> {
    const response: ApiResponse<any> = await this.apisauce.get(url, {}, { baseURL: "" })
    
    if (!response.ok) {
      const problem = getGeneralApiProblem(response)
      if (problem) return problem
    }

    return { kind: "ok", places: response.data }
  }

  async getFeed(page: number) {
    const response: ApiResponse<any> = await this.apisauce.get(`https://blogs.nasa.gov/spacestation/feed?paged=${page}`, {}, { baseURL: "" })
    
    if (!response.ok) {
      const problem = getGeneralApiProblem(response)
      if (problem) return problem
    }

    return { kind: "ok", places: response.data }
  }
}

// Singleton instance of the API for convenience
export const api = new Api()
