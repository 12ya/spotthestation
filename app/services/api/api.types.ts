/**
 * These types indicate the shape of the data you expect to receive from your
 * API endpoint, assuming it's a JSON object like we have.
 */
export interface EpisodeItem {
  title: string
  pubDate: string
  link: string
  guid: string
  author: string
  thumbnail: string
  description: string
  content: string
  enclosure: {
    link: string
    type: string
    length: number
    duration: number
    rating: { scheme: string; value: string }
  }
  categories: string[]
}

export interface ApiFeedResponse {
  status: string
  feed: {
    url: string
    title: string
    link: string
    author: string
    description: string
    image: string
  }
  items: EpisodeItem[]
}

export interface GetISSSightingsParams {
  zone: string,
  lat: number,
  lon: number
}

export interface GetISSDataParams {
  lat: number,
  lon: number
}

export interface ISSSighting {
  date: string
  maxHeight: number
  appears: string
  disappears: string
  visible: number
  dayStage: number
  notify?: boolean
}

export interface OrbitPoint {
  date: string
  latitude: number
  longitude: number,
  azimuth: number,
  elevation: number,
}

export interface ISSDataResponse {
  ok: boolean
  data: OrbitPoint[] | string
}

export interface ISSSightingResponse {
  ok: boolean
  data: ISSSighting[] | string
}

/**
 * The options used to configure apisauce.
 */
export interface ApiConfig {
  /**
   * The URL of the api.
   */
  url: string

  /**
   * Milliseconds before we timeout the request.
   */
  timeout: number
}
