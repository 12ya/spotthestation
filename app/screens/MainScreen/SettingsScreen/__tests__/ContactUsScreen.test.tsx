import React, { Component } from "react"
import { ContactUsScreen } from "../ContactUsScreen"
import { useNavigation } from "@react-navigation/native"
import { act, fireEvent, render } from "@testing-library/react-native"
import { api } from "../../../../services/api"
import { jest } from "@jest/globals"

jest.mock("@react-navigation/native", () => ({
  useNavigation: jest.fn(),
}))

jest.mock("mobx-react-lite", () => ({
  observer: (component: Component) => component,
}))

jest.mock("react-native-safe-area-context", () => ({
  useSafeAreaInsets: () => ({ top: 10 }),
}))

jest.mock("react-native-snackbar", () => ({
  show: jest.fn(),
  dismiss: jest.fn(),
}))

describe("ContactUsScreen", () => {
  const mockNavigation = {
    goBack: jest.fn(),
  }

  beforeEach(() => {
    jest.mocked(useNavigation).mockReturnValue(mockNavigation)
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  test("renders header correctly", () => {
    const { getByText } = render(<ContactUsScreen />)
    const header = getByText("settings.contactUsData.title undefined")
    expect(header).toBeTruthy()
  })

  test("navigates back when back button is pressed", () => {
    const { getByText } = render(<ContactUsScreen />)
    const backButton = getByText("settings.contactUsData.backButton undefined")
    fireEvent.press(backButton)
    expect(mockNavigation.goBack).toHaveBeenCalled()
  })

  test("sends mail when send button is pressed", async () => {
    const { getByText, getByLabelText } = render(<ContactUsScreen />)
    const sendButton = getByText("settings.contactUsData.sendButton undefined")
    const titleInput = getByLabelText("title select")
    const commentsInput = getByLabelText("comments input")

    fireEvent.changeText(titleInput, "Report an Issue")
    fireEvent.changeText(commentsInput, "This is a test comment")

    await act(async () => {
      fireEvent.press(sendButton)
      await Promise.resolve()
    })

    expect(api.sendMail).not.toHaveBeenCalled()
  })
})
