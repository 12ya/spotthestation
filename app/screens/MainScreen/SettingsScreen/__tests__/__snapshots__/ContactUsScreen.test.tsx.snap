// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <RCTScrollView
    accessibilityHint="Contact scrollable us area"
    accessibilityLabel="Contact scrollable us area"
    accessibilityRole="scrollbar"
    accessible={true}
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "paddingBottom": 86,
      }
    }
    scrollEnabled={true}
    style={
      Object {
        "paddingHorizontal": 26,
      }
    }
  >
    <View>
      <View
        accessibilityHint="Navigates to the previous screen"
        accessibilityLabel="Back button"
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "paddingBottom": 16,
          }
        }
      >
        <div />
        <div />
      </View>
      <div />
      <View
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "#1E2022",
            "borderRadius": 40,
            "height": 80,
            "justifyContent": "center",
            "marginTop": 26,
            "overflow": "hidden",
          }
        }
      >
        <View
          accessibilityLabel="title select"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 35,
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "flex": 1,
                  "fontSize": 16,
                },
                Array [
                  Object {
                    "color": "#DEDDDE",
                    "flex": 1,
                    "fontFamily": "Helvetica Neue",
                    "fontSize": 26,
                    "marginHorizontal": 17,
                    "paddingHorizontal": 0,
                    "paddingVertical": 0,
                    "textAlignVertical": "center",
                  },
                  Object {
                    "color": "#ADADAE",
                  },
                ],
                Object {},
              ]
            }
          >
            settings.contactUsData.titlePlaceholder undefined
          </Text>
          <div />
        </View>
      </View>
      <div />
      <div />
    </View>
  </RCTScrollView>
</div>
`;
